### PR TITLE
Various improvements to the Dockerfile

### PR DESCRIPTION
**Pull request by DanReyLop**
Yesterday I needed to install calypso in a local Docker container and I entered in a world of pain. I've implemented these fixes:

* Prevent the build from failing if there is no `npmrc` file.
* Prevent `npm install` from failing when running the build for the first time in a clean VM.
* And the original reason why I needed to install using Docker: copy the `npm_shrinkwrap.json` file to the container before installing the dependencies. Before this fix, we are effectively ignoring `shrinkwrap` in all the builds that use Docker (which can lead to unexplainable bugs).

@blowery @gwwar 

Note: with this PR, https://github.com/Automattic/wp-calypso/pull/4373 is no longer needed.

### CODE REVIEW

**Comment by DanReyLop** [comment]
Also, I've tried using `Vagrant` and I couldn't get it to work. I ended up just installing [Docker Toolbox](https://docs.docker.com/engine/installation/mac/) and it worked at the first try, plus is the Docker's recommended way, so maybe we should consider if we should use Vagrant at all (and of course, update the docs).

File: Dockerfile [hunk 1]
```diff
@@ -18,18 +18,23 @@ ENV NODE_VERSION 4.3.0
...
 RUN     wget https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64.tar.gz && \
           tar -zxf node-v$NODE_VERSION-linux-x64.tar.gz -C /usr/local && \
-          ln -sf node-v$NODE_VERSION-linux-x64 /usr/local/node && \
+          ln -sf /usr/local/node-v$NODE_VERSION-linux-x64 /usr/local/node && \
           ln -sf /usr/local/node/bin/npm /usr/local/bin/ && \
           ln -sf /usr/local/node/bin/node /usr/local/bin/ && \
           rm node-v$NODE_VERSION-linux-x64.tar.gz
 
-RUN     mkdir /usr/local/node/etc && cp /usr/local/etc/npmrc /usr/local/node/etc/npmrc
+RUN     touch /usr/local/etc/npmrc && \
```
Comment (gwwar) [inline Dockerfile]: A comment here would help to explain the need to create npmrc locally.

File: Dockerfile [hunk 2]
```diff
@@ -18,18 +18,23 @@ ENV NODE_VERSION 4.3.0
...
 
 ENV     NODE_PATH /calypso/server:/calypso/client
 
 # Install base npm packages to take advantage of the docker cache
 COPY    ./package.json /calypso/package.json
-RUN     npm install --production
+COPY    ./npm-shrinkwrap.json /calypso/npm-shrinkwrap.json
+
+# Sometimes "npm install" fails the first time when the cache is empty, so we retry once if it failed
+RUN     npm install --production || npm install --production
```
Comment (gwwar) [inline Dockerfile]: We'll see if we can get this fixed before this lands.

**Comment by gwwar** [comment]
:+1: Thanks for tackling this! The image builds properly here using docker toolbox and the dockerfile looks reasonable. 

I'm still sorting out what commands I need to run this image and power wpcalypso locally. So hold off on merging until someone tests that.

For anyone also looking at this, the following may be helpful:
https://docs.docker.com/engine/installation/mac/

File: Dockerfile (hunk 1 above)
Comment (DanReyLop) [inline Dockerfile]: Done

File: Dockerfile (hunk 2 above)
Comment (gwwar) [inline Dockerfile]: I couldn't reproduce the cache error in master. Could you try rebasing and removing the second try here?

Comment (DanReyLop) [inline Dockerfile]: I couldn't reproduce either, I've removed it. I'm glad I could erase that horrible hack :)

File: Dockerfile [hunk 3]
```diff
@@ -18,18 +18,21 @@ ENV NODE_VERSION 4.3.0
...
+RUN     touch /usr/local/etc/npmrc && \
+          mkdir /usr/local/node/etc && \
+          cp /usr/local/etc/npmrc /usr/local/node/etc/npmrc
 
 ENV     NODE_PATH /calypso/server:/calypso/client
 
 # Install base npm packages to take advantage of the docker cache
 COPY    ./package.json /calypso/package.json
-RUN     npm install --production
+COPY    ./npm-shrinkwrap.json /calypso/npm-shrinkwrap.json
```
Comment (gwwar) [inline Dockerfile]: We still need the `npm install` :smile:

**Comment by gwwar** [comment]
Ok, I finally got this working locally. I'm unfortunately still hitting the cache bug in docker, but I can't seem to repro in master. We can try moving forward with just the single install, unless we hit this in prod. 

This should be good to go once you re-add the `npm install --production` that was accidentally blown away. Ping me again when you've fixed that up.

**Comment by DanReyLop** [comment]
WTF, good catch, Git is weird sometimes :D

If you have the cache bug, no matter if you can't always reproduce it, I would go for the 2 `npm install`. That's a nasty hack, but to be fair it doesn't have any downsides even if things work at the first try.

I've re-commited the 2 `npm install` line.

**Comment by gwwar** [comment]
Ready to :ship: after you squash changes in this branch.

**Comment by blowery** [comment]
fantastic

**Pull request merged**